src/main/scala/io/iohk/ethereum/vm/VM.scala

```diff
@@ -13,9 +14,9 @@ object VM {
     * @param context context to be executed
     * @return result of the execution
     */
-  def run(context: ProgramContext): ProgramResult = {
-    val finalState = run(ProgramState(context))
-    ProgramResult(finalState.returnData, finalState.storage, finalState.error)
+  def run(env: ExecEnv, storage: Storage): ProgramResult = {
```

Comment:
Firstly, the `Storage` class is just a representation and the actual contract's storage will be updated after the program finishes (changes to `Storage` class will be "replayed" to actual storage)

Secondly, `Storage` is a function of contract's address, and therefore a function of `ExecEnv`. So it could be evaluated within `VM.run` or as a _body def_ within `ExecEnv`.

Finally, I would like to think `VM` as a simple function of `input => result`. And it's easier to reason when the input is a single entity. Otherwise you may find you'll need to keep adding stuff to `VM.run` to access other accounts' code, balance etc.